sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -858,14 +858,19 @@ object TypeCoercion {
         b.withNewChildren(Seq(Cast(l, CalendarIntervalType), r))
 
       case Add(l @ DateType(), r @ IntegerType()) => DateAdd(l, r)
+      case Add(l @ DateType(), r @ NullType()) => DateAdd(l, Cast(r, IntegerType))
       case Add(l @ IntegerType(), r @ DateType()) => DateAdd(r, l)
+      case Add(l @ NullType(), r @ DateType()) => DateAdd(r, Cast(l, IntegerType))
       case Subtract(l @ DateType(), r @ IntegerType()) => DateSub(l, r)
+      case Subtract(l @ DateType(), r @ NullType()) => DateSub(l, Cast(r, IntegerType))
       case Subtract(l @ DateType(), r @ DateType()) =>
```

Comment:
Can we merge the multiple rule above into one like this?
```
      case b @ BinaryOperator(l @ DateType(), r @ NullType()) =>
        b.withNewChildren(Seq(l, Cast(r, IntegerType)))
```

Reply:
If so, we might leave a trivial bug here if we set `spark.sql.optimizer.maxIterations=1`, it will not be transformed to `DateAdd `

Reply:
hmm..., I personally think that behaivour looks a little weired to me. Probably, the root cause is that `Subtract(l @ DateType(), r @ NullType()).checkInputDataTypes.isSuccess` returns true. To fix this issue, we might need to modify that check code to return false. cc: @cloud-fan

Reply:
FYI, `add` with numeric type and null type is also handled in `TypeCoercion` too